README.md
# monad-chal


ChangeLog.md
# Changelog for monad-chal

## Unreleased changes


test/Spec.hs

import qualified MCPrelude as MC

import HexDecode
import Set1
import qualified Set2 as S2
import Set3

main :: IO ()
main = do
    print ""
    print $ hexDecode "67656E6572616C41203A3A202861202D3E206229202D3E2047656E2061202D3E2047656E2062"
    print $ hexDecode "6D6B47656E203A3A2061202D3E2047656E2061"
    print $ hexDecode "64617461204D617962652061203D204E6F7468696E67207C204A7573742061"
    print $ hexDecode "202073686F77204E6F7468696E67203D20224E6F7468696E6722DA202073686F7720284A757374206129203D20224A7573742022202B2B2073686F772061"
    print $ hexDecode "794C696E6B203A3A202861202D3E2062202D3E206329202D3E204D617962652061202D3E204D617962652062202D3E204D617962652063"
    print $ hexDecode "7472616E734D61796265203A3A202861202D3E206229202D3E204D617962652061202D3E204D617962652062"
    print $ hexDecode "7461696C4D6178203A3A204F72642061203D3E205B615D202D3E204D6179626520284D61796265206129"
    print $ hexDecode "636F6D62696E65203A3A204D6179626520284D61796265206129202D3E204D617962652061"
    print $ hexDecode "636F6D6253746570203A3A205B61202D3E20625D202D3E205B615D202D3E205B625D"

    print "--------- Set 1 -----------"
    print $ fiveRands
    print $ product fiveRands == 8681089573064486461641871805074254223660
    print $ randLetter $ MC.mkSeed 4
    print $ randString3
    let
        a = fst $ MC.rand $ MC.mkSeed 1
        b = fst $ randEven $ MC.mkSeed 1
        c = fst $ randOdd $ MC.mkSeed 1
        d = fst $ randTen $ MC.mkSeed 1
    print $ product [b, c, d] == 189908109902700
    print $ fst (randPair $ MC.mkSeed 1) == ('l',282475249)
    print $ fst (generalPair randLetter MC.rand $ MC.mkSeed 1) == ('l',282475249)
    print $ fst (generalPair2 randLetter MC.rand $ MC.mkSeed 1) == ('l',282475249)
    print $ fst (repRandom (replicate 3 randLetter) $ MC.mkSeed 1) == randString3
    print $ mkGen 42 $ MC.mkSeed 1

    print "--------- Set 2 -----------"

    print $ S2.Just 5
    print $ (S2.Nothing :: S2.Maybe Int)

    print $ (S2.headMay [] :: S2.Maybe Int)
    print $ S2.headMay [1, 2]
    print $ S2.headMay [1]

    print $ (S2.tailMay [] :: S2.Maybe [Int])
    print $ S2.tailMay [1,2]
    print $ S2.tailMay [1]

    print $ (S2.lookupMay 42 [] :: S2.Maybe Int)
    print $ S2.lookupMay 42 [(42, 1)]
    print $ S2.lookupMay 42 [(1, 2)]
    print $ S2.lookupMay 42 [(1, 2), (42, 3)]

    print $ S2.divMay 4 2
    print $ S2.divMay 108 12
    print $ S2.divMay 2 4
    print $ S2.divMay 42 0

    print $ (S2.maximumMay [] :: S2.Maybe Int)
    print $ S2.maximumMay [1]
    print $ S2.maximumMay [42, 3]
    print $ S2.maximumMay [3, 108, 12]
    print $ S2.maximumMay [3, 108, 12, 108]

    print $ (S2.minimumMay [] :: S2.Maybe Int)
    print $ S2.minimumMay [1]
    print $ S2.minimumMay [42, 3]
    print $ S2.minimumMay [108, 3, 12]
    print $ S2.minimumMay [108, 3, 12, 3]

    print $ MC.greekDataA
    print $ MC.greekDataB

    print $ S2.queryGreek MC.greekDataA "beta"

    print $ S2.queryGreek MC.greekDataA "alpha" == S2.Just 2.0
    print $ S2.queryGreek MC.greekDataA "beta" == S2.Nothing
    print $ S2.queryGreek MC.greekDataA "gamma" == S2.Just 3.3333333333333335
    print $ S2.queryGreek MC.greekDataA "delta" == S2.Nothing
    print $ S2.queryGreek MC.greekDataA "zeta" == S2.Nothing

    print $ S2.queryGreek MC.greekDataB "rho" == S2.Nothing
    print $ S2.queryGreek MC.greekDataB "phi" == S2.Just 0.24528301886792453
    print $ S2.queryGreek MC.greekDataB "chi" == S2.Just 9.095238095238095
    print $ S2.queryGreek MC.greekDataB "psi" == S2.Nothing
    print $ S2.queryGreek MC.greekDataB "omega" == S2.Just 24.0

    print $ S2.queryGreek2 MC.greekDataA "beta"

    print $ S2.queryGreek2 MC.greekDataA "alpha" == S2.Just 2.0
    print $ S2.queryGreek2 MC.greekDataA "beta" == S2.Nothing
    print $ S2.queryGreek2 MC.greekDataA "gamma" == S2.Just 3.3333333333333335
    print $ S2.queryGreek2 MC.greekDataA "delta" == S2.Nothing
    print $ S2.queryGreek2 MC.greekDataA "zeta" == S2.Nothing
    print $ S2.queryGreek2 MC.greekDataB "rho" == S2.Nothing
    print $ S2.queryGreek2 MC.greekDataB "phi" == S2.Just 0.24528301886792453
    print $ S2.queryGreek2 MC.greekDataB "chi" == S2.Just 9.095238095238095
    print $ S2.queryGreek2 MC.greekDataB "psi" == S2.Nothing
    print $ S2.queryGreek2 MC.greekDataB "omega" == S2.Just 24.0
    print $ S2.addSalaries salaries "bob" "john"
    print $ S2.addSalaries2 salaries "alice" "carol"
    print $ S2.addSalaries2 salaries "bob" "john"
    print $ S2.yLink (+) (S2.Just 1) (S2.Just 5)
    print $ S2.tailProd [1, 2, 3]
    print $ S2.tailProd [1, 2]
    print $ S2.tailProd [1]
    print $ S2.tailProd []
    print $ S2.tailSum [1, 2, 3]
    print $ S2.tailSum [1, 2]
    print $ S2.tailSum [1]
    print $ S2.tailSum []
    print $ S2.combine (S2.Just (S2.Just 1))
    print $ S2.tailMax [1, 3, 42]
    print $ S2.tailMax [1000, 42]
    print $ S2.tailMax [1000]
    print $ (S2.tailMax [] :: S2.Maybe Int)
    print $ S2.tailMin [5, 42, 1]
    print $ S2.tailMin [1, 42]
    print $ S2.tailMin [1000]
    print $ (S2.tailMin [] :: S2.Maybe Int)
    
    print "--------- Set 3 -----------"
    print $ allPairs1 [1,2] [3,4]
    print $ allPairs1 [1,2,3] [4,5,6]
    print $ allPairs1 [1..3] [6..8]
    print $ allPairs1 [1..3] [6..8] == [(1,6),(1,7),(1,8),(2,6),(2,7),(2,8),(3,6),(3,7),(3,8)]
    print $ allPairs [1,2] [3,4]
    print $ allPairs [1,2,3] [4,5,6]
    print $ allPairs [1..3] [6..8]
    print $ allPairs [1..3] [6..8] == [(1,6),(1,7),(1,8),(2,6),(2,7),(2,8),(3,6),(3,7),(3,8)]
    print $ MC.cardRanks
    print $ MC.cardSuits
    print $ allPairs1 MC.cardRanks MC.cardSuits
    print $ allPairs MC.cardRanks MC.cardSuits
    print $ Card 1 "H"
    print $ show (Card 2 "h") == "2h"
    print $ show (allCards1 MC.cardRanks MC.cardSuits) == "[2H,2D,2C,2S,3H,3D,3C,3S,4H,4D,4C,4S,5H,5D,5C,5S]"
    print $ show (allCards MC.cardRanks MC.cardSuits) == "[2H,2D,2C,2S,3H,3D,3C,3S,4H,4D,4C,4S,5H,5D,5C,5S]"
    print $ show (allCards1 MC.cardRanks MC.cardSuits)
    print $ allCombs35 (,,) [1,2] [3,4] [5,6] == [(1,3,5),(1,3,6),(1,4,5),(1,4,6),(2,3,5),(2,3,6),(2,4,5),(2,4,6)]

salaries :: [(String, Integer)]
salaries = [ ("alice", 105000)
           , ("bob", 90000)
           , ("carol", 85000)
           ]


MonadChallenge.hsproj/Set1.hs
{-# LANGUAGE MonadComprehensions #-}
{-# LANGUAGE RebindableSyntax  #-}

module Set1 where
  
import MCPrelude

-- required for Haskell for Mac to display values in playground
import Prelude (($!)) 

-----------------------------------------------
-- 1. Random Number Generation
-----------------------------------------------
fiveRands :: [Integer]
fiveRands = take 5 $ drop 1 $ map fst $ iterate (\(_, seed) -> rand seed) (0, mkSeed 1)
    
type Gen a = Seed -> (a, Seed)

-----------------------------------------------
-- 2 Random Character Generation
-----------------------------------------------
randLetter :: Gen Char
randLetter = \seed -> case rand seed of
  (n , seed) -> (toLetter n, seed)

randString3 :: String
randString3 = take 3 $ drop 1 $ map fst  $
  iterate (\(_, seed) -> randLetter seed) ('0', mkSeed 1)

-----------------------------------------------
-- 3. More Generators
-----------------------------------------------
randEven :: Gen Integer -- the output of rand * 2
randEven = generalA (* 2) rand

randOdd :: Gen Integer -- the output of rand * 2 + 1
randOdd = generalA ((+ 1) . (* 2)) rand

randTen :: Gen Integer -- the output of rand * 10
randTen = generalA (* 10) rand
  
generalA :: (a -> b) -> Gen a -> Gen b
generalA a2b genA = \seed -> case genA seed of
  (a, sd) -> (a2b a, sd)

-----------------------------------------------
-- 4. Generalizing Random Pairs
-----------------------------------------------
randPair :: Gen (Char, Integer)
randPair = \seed ->
  let
    (ch, chSeed) = (randLetter seed)
    (int, newSeed) = rand chSeed
  in
    ((ch, int), newSeed)
    
generalPair :: Gen a -> Gen b -> Gen (a,b)
generalPair genA genB = \seed ->
  let
    (a, aSeed) = genA seed
    (b, newSeed) = genB aSeed
  in
    ((a, b), newSeed)
    
generalB :: (a -> b -> c) -> Gen a -> Gen b -> Gen c
generalB ab2c genA genB = \seed ->
  let
    (a, aSeed) = genA seed
    (b, bSeed) = genB aSeed
    (c, newSeed) = ((ab2c a b), bSeed)
  in
    (c, newSeed)
    
generalPair2:: Gen a -> Gen b -> Gen (a,b)
generalPair2 = generalB (,)

-----------------------------------------------
-- 5. Generalizing Lists of Generators
-----------------------------------------------
repRandom :: [Gen a] -> Gen [a]
--repRandom :: [Seed -> (a, Seed)] -> (Seed -> ([a], Seed))
repRandom genAs = \seed0 ->
  let
    f (as, seed) genA  = case genA seed of
      (a, sd) -> (a : as, sd)
  in
    case foldl f ([], seed0) genAs of
      (as, resultSeed) -> (reverse as, resultSeed)
      
-----------------------------------------------    
-- 6. Threading the random number state
-----------------------------------------------
genTwo :: Gen a -> (a -> Gen b) -> Gen b
genTwo genA k = \seed -> case genA seed of
  (a, sd) -> k a sd
  
mkGen :: a -> Gen a
mkGen a = \seed -> (a, seed)


MonadChallenge.hsproj/Set2.hs
{-# LANGUAGE MonadComprehensions #-}
{-# LANGUAGE RebindableSyntax  #-}

module Set2 where
  
import MCPrelude

-- required for Haskell for Mac to display values in playground
import Prelude (($!)) 

-----------------------------------------------
-- 1. The Maybe Type
-----------------------------------------------

data Maybe a = Just a | Nothing
  deriving (Eq, Show) 

{-- the exercise wants us to do this, but "deriving" is easier
 -- the exercise should have had us define Eq as well
 
instance Show a => Show (Maybe a) where
  show (Just a)= "Just " ++ show a
  show Nothing = "Nothing"
--}

-----------------------------------------------
-- 2. Build a library of things that can fail
-----------------------------------------------

headMay :: [a] -> Maybe a
headMay [] = Nothing
headMay (a : _) = Just a

tailMay :: [a] -> Maybe [a]
tailMay [] = Nothing
tailMay (_ : as) = Just as

lookupMay :: Eq a => a -> [(a, b)] -> Maybe b
lookupMay a [] = Nothing
lookupMay a' ((a, b) : tups) =
  ifThenElse (a' == a) (Just b) $ lookupMay a' tups

divMay :: (Eq a, Fractional a) => a -> a -> Maybe a
divMay _ 0 = Nothing
divMay a1 a2 = Just (a1 / a2)

maximumMay :: Ord a => [a] -> Maybe a
maximumMay [] = Nothing
maximumMay [a] = Just a
maximumMay (a : as) = case maximumMay as of
   Just m -> Just $ a `max` m
   Nothing -> Just a

minimumMay :: Ord a => [a] -> Maybe a
minimumMay [] = Nothing
minimumMay [a] = Just a
minimumMay (a : as) = case minimumMay as of
   Just m -> Just $ a `min` m
   Nothing -> Just a
   
-----------------------------------------------
-- 3. Chains of Failing Computations
-----------------------------------------------
queryGreek :: GreekData -> String -> Maybe Double
queryGreek greekData key =
  case lookupMay key greekData of
      Nothing -> Nothing
      Just xs -> case tailMay xs of
        Nothing -> Nothing
        Just ys -> case maximumMay ys of
          Nothing -> Nothing
          Just mx -> case headMay xs of
            Nothing -> Nothing
            Just first -> divMay (fromIntegral mx) (fromIntegral first)

-----------------------------------------------
-- 4. Generalizing chains of failures
-----------------------------------------------

chain :: (a -> Maybe b) -> Maybe a -> Maybe b
chain = flip link

link :: Maybe a -> (a -> Maybe b) -> Maybe b
link Nothing f = Nothing
link (Just a) f = f a

-- I have an idea, let's just call link ">>=" and use infix notation
(>>=) = link

queryGreek2 :: GreekData -> String -> Maybe Double
queryGreek2 greekData key =
  let
    xsMaybe = lookupMay key greekData
    ysMaybe = xsMaybe >>= tailMay
    mxMaybe = ysMaybe >>= maximumMay
    firstMaybe = xsMaybe >>= headMay
  in
    firstMaybe >>= (\first ->
      mxMaybe >>= (\mx ->
        divMay (fromIntegral mx) (fromIntegral first)))
        
-----------------------------------------------
-- 5. Chaining variations
-----------------------------------------------
addSalaries :: [(String, Integer)] -> String -> String -> Maybe Integer
addSalaries salaries name1 name2 =
  let
    maybeS1 = lookupMay name1 salaries
    maybeS2 = lookupMay name2 salaries
  in
    maybeS1 >>= \s1 ->
    maybeS2 >>= \s2 ->
      mkMaybe $ s1 + s2
      
-- in other words, "liftM2"
yLink :: (a -> b -> c) -> Maybe a -> Maybe b -> Maybe c
yLink f maybeA maybeB =
  maybeA >>= \a ->
  maybeB >>= \b ->
    mkMaybe $ f a b
    
addSalaries2 :: [(String, Integer)] -> String -> String -> Maybe Integer
addSalaries2 salaries name1 name2 =
  let
    maybeS1 = lookupMay name1 salaries
    maybeS2 = lookupMay name2 salaries
  in
    yLink (+) maybeS1 maybeS2
    
-- In other words, "return" or "pure"
mkMaybe :: a -> Maybe a
mkMaybe = Just
        
-----------------------------------------------
-- 6. Tailprod
-----------------------------------------------
tailProd :: Num a => [a] -> Maybe a
--tailProd xs = tailMay xs >>= \as -> Just $ product as
tailProd = transMaybe product . tailMay

  
tailSum :: Num a => [a] -> Maybe a
--tailSum xs = tailMay xs >>= \as -> Just $ sum as
tailSum = transMaybe sum . tailMay
  
-- in other words, "fmap"
transMaybe :: (a -> b) -> Maybe a -> Maybe b
transMaybe f maybeA = maybeA >>= \a -> mkMaybe $ f a

--tailMax :: Ord a => [a] -> Maybe (Maybe a)
tailMax :: Ord a => [a] -> Maybe a
tailMax = combine . transMaybe maximumMay . tailMay

tailMin :: Ord a => [a] -> Maybe a
tailMin = combine . transMaybe minimumMay . tailMay

-- in other words, "join"
combine :: Maybe (Maybe a) -> Maybe a
combine mma = mma >>= id


MonadChallenge.hsproj/Set3.hs
{-# LANGUAGE MonadComprehensions #-}
{-# LANGUAGE RebindableSyntax  #-}

module Set3 where
  
import MCPrelude

-- required for Haskell for Mac to display values in playground
import Prelude (($!)) 

-- 1. Generating combinations
allPairs1 :: [a] -> [b] -> [(a,b)]
allPairs1 as bs =
  let
    nextPair [] _ = []
    nextPair (_ : xs) [] = nextPair xs bs
    nextPair xs@(x : _) (y : ys) = (x, y) : nextPair xs ys
  in
    nextPair as bs
    
    
-- 2. Poker hands
data Card = Card Int String

instance Show (Card) where
  show (Card a b)= show a ++ b
  
allCards1 :: [Int] -> [String] -> [Card]
allCards1 ranks suits =
  let
    nextCard [] _ = []
    nextCard (_ : rs) [] = nextCard rs suits
    nextCard rs@(r : _) (s : ss) = Card r s : nextCard rs ss
  in
    nextCard ranks suits
   

-- 3. Generalizing pairs and cards
allCombs1 :: (a -> b -> c) -> [a] -> [b] -> [c]
allCombs1 f as bs =
  let
    next [] _ = []
    next (_ : xs) [] = next xs bs
    next xs@(x : _) (y : ys) = f x y : next xs ys
  in
    next as bs


allPairs :: [a] -> [b] -> [(a,b)]
allPairs = allCombs (,)

allCards :: [Int] -> [String] -> [Card]
allCards = allCombs Card

-- 4. Combinations of three things
allCombs34 :: (a -> b -> c -> d) -> [a] -> [b] -> [c] -> [d]
allCombs34 f as bs cs =
  let
    next [] _ _ = []
    next (_ : xs) [] _ = next xs bs cs
    next xs (_ : ys) [] = next xs ys cs
    next xs@(x : _) ys@(y : _) (z : zs) = f x y z : next xs ys zs
  in
    next as bs cs

-- 5. Combinations of more things
combStep :: [a -> b] -> [a] -> [b]
combStep fs as =
  let
    next [] _ = []
    next (_ : fs) [] = next fs as
    next fs@(f : _) (y : ys) = f y : next fs ys
  in
    next fs as
      
allCombs5 :: (a -> b -> c) -> [a] -> [b] -> [c]
allCombs5 f as bs =
  combStep (combStep [f] as) bs
  
allCombs35 :: (a -> b -> c -> d) -> [a] -> [b] -> [c] -> [d]
allCombs35 f as bs cs =
  combStep (combStep (combStep [f] as) bs) cs
  
-- hey, what if we use infix and call "combStep" <*> ?
(<*>) :: [a -> b] -> [a] -> [b]
(<*>) fs as =
  let
    next [] _ = []
    next (_ : fs) [] = next fs as
    next fs@(f : _) (y : ys) = f y : next fs ys
  in
    next fs as
      
allCombs :: (a -> b -> c) -> [a] -> [b] -> [c]
allCombs f as bs = [f] <*> as <*> bs
  
allCombs3 :: (a -> b -> c -> d) -> [a] -> [b] -> [c] -> [d]
allCombs3 f as bs cs = [f] <*> as <*> bs <*> cs

-- the first step could be just fmap and called <$>
-- f <$> as <*> bs <*> cs






MonadChallenge.hsproj/MCPrelude.hs
--module MCPrelude (
--    Seed,
--    mkSeed,
--    rand
--    ) where
--      
--newtype Seed = Seed { unSeed :: Integer }
--  deriving (Eq,Show)
--
--mkSeed :: Integer -> Seed
--mkSeed n = Seed n
--
--m :: Integer
--m = 0x7FFFFFFF
--
--rand :: Seed -> (Integer, Seed)
--rand (Seed s) = (s', Seed s')
--  where
--    s' = (s * 16807) `mod` m

{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE CPP, NoImplicitPrelude, BangPatterns #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  MCPrelude
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  mightybyte@gmail.com
-- Stability   :  stable
-- Portability :  portable
--
-- MCPrelude is a modified version of the Haskell Prelude designed
-- specifically for The Monad Challenges.
--
-----------------------------------------------------------------------------

module MCPrelude (

    -- * Standard types, classes and related functions

    -- ** Basic data types
    Bool(False, True),
    (&&), (||), not, otherwise,

    Ordering(LT, EQ, GT),
    Char, String,

    ifThenElse,

    -- *** Tuples
    fst, snd, curry, uncurry,

    -- ** Basic type classes
    Eq((==), (/=)),
    Ord(compare, (<), (<=), (>=), (>), max, min),
    Enum(succ, pred, toEnum, fromEnum, enumFrom, enumFromThen,
         enumFromTo, enumFromThenTo),
    Bounded(minBound, maxBound),

    -- ** Numbers

    -- *** Numeric types
    Int, Integer, Float, Double,
    Rational,

    -- *** Numeric type classes
    Num((+), (-), (*), negate, abs, signum, fromInteger),
    Real(toRational),
    Integral(quot, rem, div, mod, quotRem, divMod, toInteger),
    Fractional((/), recip, fromRational),
    Floating(pi, exp, log, sqrt, (**), logBase, sin, cos, tan,
             asin, acos, atan, sinh, cosh, tanh, asinh, acosh, atanh),
    RealFrac(properFraction, truncate, round, ceiling, floor),
    RealFloat(floatRadix, floatDigits, floatRange, decodeFloat,
              encodeFloat, exponent, significand, scaleFloat, isNaN,
              isInfinite, isDenormalized, isIEEE, isNegativeZero, atan2),

    -- *** Numeric functions
    subtract, even, odd, gcd, lcm, (^), (^^),
    fromIntegral, realToFrac,

    -- ** Miscellaneous functions
    id, const, (.), flip, ($), until,

    -- * List operations
    map, (++), filter,
    null, length, (!!),
    reverse,
    -- ** Reducing lists (folds)
    foldl, foldl1, foldr, foldr1,
    -- *** Special folds
    and, or, any, all,
    product, sum,
    concat, concatMap,
    -- ** Building lists
    -- *** Scans
    scanl, scanl1, scanr, scanr1,
    -- *** Infinite lists
    iterate, repeat, replicate, cycle,
    -- ** Sublists
    take, drop, splitAt, takeWhile, dropWhile, span, break,
    -- ** Searching lists
    elem, notElem,
    -- ** Zipping and unzipping lists
    zip, zip3, zipWith, zipWith3, unzip, unzip3,
    -- ** Functions on strings
    lines, words, unlines, unwords,

    -- * Converting to and from @String@
    -- ** Converting to @String@
    ShowS,
    Show(showsPrec, showList, show),
    shows,
    showChar, showString, showParen,
    -- ** Converting from @String@
    ReadS,
    Read(readsPrec, readList),
    reads, readParen, read, lex,
    undefined,

    Seed,
    mkSeed,
    rand,
    toLetter,

    GreekData,
    greekDataA,
    greekDataB,
    salaries,
    firstNames,
    lastNames,
    cardRanks,
    cardSuits

  ) where

import Data.Char
import Data.List
import Data.Tuple

import GHC.Base hiding (foldr)
import GHC.Enum
import GHC.Num
import GHC.Real
import GHC.Float
import GHC.Show
import Prelude (undefined)
import Text.Read

infixr 0 $!

ifThenElse :: Bool -> a -> a -> a
ifThenElse True a _ = a
ifThenElse False _ a = a

-- -----------------------------------------------------------------------------
-- Miscellaneous functions

-- | Strict (call-by-value) application, defined in terms of 'seq'.
($!)    :: (a -> b) -> a -> b
f $! x  = let !vx = x in f vx  -- see #2273

#ifdef __HADDOCK__
-- | The value of @'seq' a b@ is bottom if @a@ is bottom, and otherwise
-- equal to @b@.  'seq' is usually introduced to improve performance by
-- avoiding unneeded laziness.
seq :: a -> b -> b
seq _ y = y
#endif

newtype Seed = Seed { unSeed :: Integer }
  deriving (Eq,Show)

mkSeed :: Integer -> Seed
mkSeed n = Seed n

m :: Integer
m = 0x7FFFFFFF

rand :: Seed -> (Integer, Seed)
rand (Seed s) = (s', Seed s')
  where
    s' = (s * 16807) `mod` m

toLetter :: Integer -> Char
toLetter = chr . (ord 'a' + ) . (`mod` 26) . fromIntegral

type GreekData = [(String, [Integer])]

greekDataA :: GreekData
greekDataA = [ ("alpha", [5, 10])
             , ("beta", [0, 8])
             , ("gamma", [18, 47, 60])
             , ("delta", [42])
             ]

greekDataB :: GreekData
greekDataB = [ ("phi", [53, 13])
             , ("chi", [21, 8, 191])
             , ("psi", [])
             , ("omega", [6, 82, 144])
             ]

salaries :: [(String, Integer)]
salaries = [ ("alice", 105000)
           , ("bob", 90000)
           , ("carol", 85000)
           ]

firstNames :: [String]
firstNames = ["alice", "bob", "carol", "dave"]

lastNames :: [String]
lastNames = ["doe", "jones", "smith"]

cardRanks :: [Int]
cardRanks = [2, 3, 4, 5]

cardSuits :: [String]
cardSuits = ["H", "D", "C", "S"]
